### ISSUE TITLE
bug: calling save() on a model in "pcms shell" requires a transaction

### ISSUE THREAD

**Issue by robvdl** [robvdl/pyramidcms#16]
At the moment, creating a new object and calling .save() on a model in the "pcms shell" will appear to do nothing (increase sqlalchemy logging level in pyramidcms.ini to verify this that there are no queries going to the db)

This is because you need to do manually set up a transaction, by importing "transaction" and doing something like this:

```
with transaction.manager:
    obj.save()
```

It would be nice however, to not have to do this in the shell itself, as the shell is for experimentation, so you want it to run queries instantly.

I am not quite sure how we are going to be fixing this just yet.

In normal requests, we use a library "pyramid_tm" which automatically runs each request in a single transaction.

However, a management command is a bit more low level an we need to manually set up the transaction (see loaddata command for example).

Somehow, we need to get it to create a new transaction and commit it each time when calling .save(), but we only want this behaviour in the shell and not anywhere else.  Because in standard requests we want to continue using pyramid_tm>

**Comment by robvdl** [comment]
John: If you are going to work on this, check if SQL Alchemy has something that will autocommit each time (some sort of global autocommit setting that we turn on in the shell command only), I would rather have this than putting extra code in the save() method of each model.

**Comment by robvdl** [comment]
We should be able to do this by setting autocommit=True here:

    DBSession = scoped_session(sessionmaker(extension=ZopeTransactionExtension(), autocommit=True))

However, the problem here is we only want to do this for the "pcms shell" management command, so I need to have a think how we are going to do this.

Then the other problem is that I am getting exceptions when inserting records into the database in the pcms shell with autocommit=True:

    >>> from pyramidcms.db.models import *
    >>> u = User(username='foo')
    >>> u.set_password('password')
    >>> u.save()
    AttributeError: 'NoneType' object has no attribute '_iterate_parents' (????? what ???)

I also tried taking out the extension=ZopeTransactionExtension() part and left autocommit=True on, however that did something strange too, creating a user "u.save()" did nothing (turn on SQLAlchemy logging in pyramidcms.ini first to see), however doing the first query to get a list of users "User.objects.all()" triggered the insert query to be executed which is not correct.. it should have already done this when calling user.save()

**Comment by robvdl** [comment]
Another thing I tried, following the SQLAlchemy docs is to also set autoflush=False and expire_on_commit=False together with autocommit=True when creating the DBSession, and not using ZopeTransactionExtension this time.

However, same problem as described above, if you don't use ZopeTransactionExtension, creating a simple user and calling save() won't do anything... until you do a query to return a list of users which is wrong.

**Comment by robvdl** [comment]
John: I unassigned you because this clearly is a more complex problem than I thought, going to leave it open for now and focus on more important things as this only affects the shell.

**Comment by robvdl** [comment]
We should have probably looked at hooking into this for our command framework, I wasn't aware of this at the time:

http://docs.pylonsproject.org/docs/pyramid/en/latest/narr/commandline.html#writing-a-script

Not sure if it will make much of a difference here though.

**Comment by johndanielcurley** [comment]
I will have a look at this and try to integrate with pyramid's native shell.

**Comment by johndanielcurley** [comment]
I have created a prototype shell command. I think we could add the model session to the ini file that is allowed in custom shells.

**Comment by robvdl** [comment]
that is interesting, I will have a read of this later.

in the mean time, your work on this branch has been merged with the stuff from the scaffold branch, into master.

**Comment by johndanielcurley** [comment]
I have added a commit() command to the model manager. While this isn't as elegant as having a save() that works within the shell, I do not think this is possible due to the nature of pyramid_tm running after all transactions are completed.

Other options I have looked at include:

Using sqlalchemy inspect to check whether an operation is pending or not. However I do not think this will work as the pyramid_tm works after all the methods have completed.

Having an optional arg/kwarg "shell" that lets the save() command know that we're dealing with the shell. However this just seems awkward and less elegant than the "commit() command.
